test/lms_test.c

```diff
@@ -0,0 +1,538 @@
+    ret = 1;
+err:
+    EVP_PKEY_free(pub);
+    OSSL_DECODER_CTX_free(dctx);
+    return ret;
+}
+
+static int lms_key_eq_test(void)
+{
+    int ret = 0;
+    EVP_PKEY *key[4];
+    LMS_ACVP_TEST_DATA *td1 = &lms_testdata[0];
+    LMS_ACVP_TEST_DATA *td2 = &lms_testdata[1];
+    size_t i;
+#ifndef OPENSSL_NO_EC
+    EVP_PKEY *eckey = NULL;
+#endif
+
+    for (i = 0; i < OSSL_NELEM(key); i++)
+        key[i] = NULL;
+
+    if (!TEST_ptr(key[0] = lms_pubkey_from_data(td1->pub, td1->publen))
+            || !TEST_ptr(key[1] = lms_pubkey_from_data(td1->pub, td1->publen))
+            || !TEST_ptr(key[2] = lms_pubkey_from_data(td2->pub, td2->publen))
+            || !TEST_ptr(key[3] = key_decode_from_data(td1->pub, td1->publen,
+                                                       NULL)))
+        goto end;
+
+    ret = TEST_int_eq(EVP_PKEY_eq(key[0], key[1]), 1)
+        && TEST_int_ne(EVP_PKEY_eq(key[0], key[2]), 1)
+        && TEST_int_eq(EVP_PKEY_eq(key[0], key[3]), 1);
+    if (ret == 0)
+        goto end;
+
+#ifndef OPENSSL_NO_EC
+    if (!TEST_ptr(eckey = EVP_PKEY_Q_keygen(libctx, NULL, "EC", "P-256")))
+        goto end;
+    ret = TEST_int_ne(EVP_PKEY_eq(key[0], eckey), 1);
+    EVP_PKEY_free(eckey);
+#endif
+end:
+    EVP_PKEY_free(key[3]);
+    EVP_PKEY_free(key[2]);
+    EVP_PKEY_free(key[1]);
+    EVP_PKEY_free(key[0]);
+    return ret;
+}
+
+static int lms_key_validate_test(void)
+{
+    int ret = 0;
+    LMS_ACVP_TEST_DATA *td = &lms_testdata[0];
+    EVP_PKEY_CTX *vctx = NULL;
+    EVP_PKEY *key = NULL;
+
+    if (!TEST_ptr(key = lms_pubkey_from_data(td->pub, td->publen)))
+        return 0;
+    if (!TEST_ptr(vctx = EVP_PKEY_CTX_new_from_pkey(libctx, key, NULL)))
+        goto end;
+    ret = TEST_int_eq(EVP_PKEY_check(vctx), 1);
+    EVP_PKEY_CTX_free(vctx);
+end:
+    EVP_PKEY_free(key);
+    return ret;
+}
+
+static int lms_verify_test(int tst)
+{
+    int ret = 0;
+    LMS_ACVP_TEST_DATA *td = &lms_testdata[tst];
+    EVP_PKEY_CTX *ctx = NULL;
+    EVP_PKEY *pkey = NULL;
+    EVP_SIGNATURE *sig = NULL;
+
+    ret = TEST_ptr(pkey = lms_pubkey_from_data(td->pub, td->publen))
+        && TEST_ptr(sig = EVP_SIGNATURE_fetch(libctx, "LMS", NULL))
+        && TEST_ptr(ctx = EVP_PKEY_CTX_new_from_pkey(libctx, pkey, NULL))
+        && TEST_int_eq(EVP_PKEY_verify_message_init(ctx, sig, NULL), 1)
+        && TEST_int_eq(EVP_PKEY_verify(ctx, td->sig, td->siglen,
+                                       td->msg, td->msglen), 1);
+
+    EVP_PKEY_free(pkey);
+    EVP_PKEY_CTX_free(ctx);
+    EVP_SIGNATURE_free(sig);
+    return ret;
+}
+
+static int lms_digest_verify_fail_test(void)
+{
+    int ret = 0;
+    LMS_ACVP_TEST_DATA *td = &lms_testdata[0];
+    EVP_PKEY *pub = NULL;
+    EVP_MD_CTX *vctx = NULL;
+
+    if (!TEST_ptr(pub = lms_pubkey_from_data(td->pub, td->publen)))
+        return 0;
+    if (!TEST_ptr(vctx = EVP_MD_CTX_new()))
+        goto err;
+    if (!TEST_int_eq(EVP_DigestVerifyInit_ex(vctx, NULL, NULL, libctx, NULL,
```

Comment:
Why can't we make this work?

Reply:
Same reason, only one shot is supported.  This is true of all the PQ signatures.